# Preprocessing

In this notebook, I'll be preprocessing all my data to get it ready to be put into a model.

In [1]:
#load in
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
pd.set_option('display.max_columns', 100)
import numpy as np


## Data Loading

Just checking to see if my data has been correctly imported.

In [2]:
#load in dataset
df=pd.read_csv("data1/my_dataset.csv")
#look at sample
df.sample(5)

player_id  last_season_x  most_recent_club_id country_of_citizenship  \
8794      107377           2018                28643                Croatia   
20847     624436           2022                 1038                  Spain   
24046       4391           2023                  383            Netherlands   
77775     181380           2023                  114               DR Congo   
70353     204049           2023                 3209                Albania   

             sub_position    position   foot  height_in_cm  market_value  \
8794   Attacking Midfield    Midfield  right         189.0       3000000   
20847  Defensive Midfield    Midfield  right         170.0        500000   
24046          Goalkeeper  Goalkeeper  right         188.0       1250000   
77775           Left-Back    Defender   left         179.0      10000000   
70353  Attacking Midfield    Midfield   both         178.0       4000000   

       highest_ever_market_value        date  age  contract_days_left  month  \
8794                   4000000.0  2018-06-03   27                  50      6   
20847                   500000.0  2023-06-15   21                 597      6   
24046                  3000000.0  2012-01-16   28                 232      1   
77775                 10000000.0  2019-09-12   26                 597      9   
70353                  5000000.0  2022-03-31   25                 597      3   

       year  goals  assists  yellow_cards  red_cards  minutes_played  \
8794   2018      1        3             4          0            1215   
20847  2023      0        0             0          0              26   
24046  2012      0        0             0          0            1947   
77775  2019      0        1             6          0            1340   
70353  2022      4        5             2          0            1521   

       player_club_id league_id  net_transfer_record  national_team_players  
8794              157       BE1            2520000.0                      8  
20847            1038       IT1            1750000.0                      4  
24046             383       NL1           -3530000.0                     10  
77775             379       GB1           18840000.0                     13  
70353            2293       TR1            1800000.0                      4

Check if datatypes have been 

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123594 entries, 0 to 123593
Data columns (total 24 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   player_id                  123594 non-null  int64  
 1   last_season_x              123594 non-null  int64  
 2   most_recent_club_id        123594 non-null  int64  
 3   country_of_citizenship     123594 non-null  object 
 4   sub_position               123594 non-null  object 
 5   position                   123594 non-null  object 
 6   foot                       123594 non-null  object 
 7   height_in_cm               123594 non-null  float64
 8   market_value               123594 non-null  int64  
 9   highest_ever_market_value  123594 non-null  float64
 10  date                       123594 non-null  object 
 11  age                        123594 non-null  int64  
 12  contract_days_left         123594 non-null  int64  
 13  month                      12

- The `player_id` collumn seems relatively useless in the case of predicting market value as its just an identifier.
- The `most_recent_club_id` collumn also seems useless as I already have the player_club_id collumn making it useless.
- We can also drop the `highest_ever_market_value` collumn as its way too correlated (see EDA notebook) with the target variable.

We can drop them both.

In [4]:
df=df.drop(["player_id","most_recent_club_id","highest_ever_market_value"],axis=1)

In [5]:
df.describe()

last_season_x   height_in_cm  market_value            age  \
count  123594.000000  123594.000000  1.235940e+05  123594.000000   
mean     2021.424948     182.265814  5.106035e+06      25.031830   
std         2.298380       6.695002  1.061903e+07       3.985102   
min      2012.000000     159.000000  1.000000e+04      15.000000   
25%      2021.000000     178.000000  5.000000e+05      22.000000   
50%      2023.000000     183.000000  1.500000e+06      25.000000   
75%      2023.000000     187.000000  4.500000e+06      28.000000   
max      2023.000000     207.000000  2.000000e+08      44.000000   

       contract_days_left          month           year          goals  \
count       123594.000000  123594.000000  123594.000000  123594.000000   
mean           443.834749       6.428346    2018.691587       1.973793   
std            471.763162       3.587796       2.976855       3.695028   
min           -314.000000       1.000000    2012.000000       0.000000   
25%            232.000000       3.000000    2017.000000       0.000000   
50%            232.000000       6.000000    2019.000000       1.000000   
75%            597.000000      10.000000    2021.000000       2.000000   
max           3154.000000      12.000000    2023.000000      59.000000   

             assists   yellow_cards      red_cards  minutes_played  \
count  123594.000000  123594.000000  123594.000000   123594.000000   
mean        1.501165       2.737706       0.068798     1308.635727   
std         2.450115       2.855257       0.265803     1010.055198   
min         0.000000       0.000000       0.000000        1.000000   
25%         0.000000       0.000000       0.000000      439.000000   
50%         1.000000       2.000000       0.000000     1143.000000   
75%         2.000000       4.000000       0.000000     1990.000000   
max        33.000000      23.000000       3.000000     5070.000000   

       player_club_id  net_transfer_record  national_team_players  
count   123594.000000         1.235940e+05          123594.000000  
mean      3175.928945         2.138397e+05               6.551669  
std       7846.777161         3.819746e+07               5.306323  
min          3.000000        -1.951000e+08               0.000000  
25%        294.000000        -1.930000e+06               2.000000  
50%        873.000000         1.500000e+05               6.000000  
75%       2457.000000         7.600000e+06              10.000000  
max      83678.000000         1.020000e+08              22.000000

Everything looks good, Lets get going with the preprocessing.

First, to avoid data leakage we will do our train test split before any proprocessing.

## Splitting

In [6]:
#Split my X and y
y=df["market_value"]
X=df.drop(["market_value"],axis=1)

In [7]:
#import splitter
from sklearn.model_selection import train_test_split

#split between train and val_test
X_train, X_val_test, y_train, y_val_test = train_test_split(X, y, test_size=0.33, random_state=100)

In [8]:
#split between test and val
X_val, X_test, y_val, y_test = train_test_split(X_val_test, y_val_test, test_size=0.5, random_state=100)

In [9]:
#check shape
X_train.shape

(82807, 20)

In [10]:
#check shape
X_val.shape

(20393, 20)

In [11]:
#check shape
X_test.shape

(20394, 20)

my splitting seems to have been correctly done. All the rows add up to the original size of the dataframe.

## Feature Encoding

Since I've got the month and year collumn i dont need the date collumn so I'll drop that.

In [12]:
X_train=X_train.drop(["date"],axis=1)
X_test=X_test.drop(["date"],axis=1)
X_val=X_val.drop(["date"],axis=1)

I need to encode the following collumns to be processed:

- `country_of_citizenship`
- `sub_position`
- `date`
- `position` 
- `foot` 
- `league_id`
  

We'll first look at the cardinality of all our categorical collumns to decide what type of encoding we should do.

In [13]:
df.nunique()

last_season_x               12
country_of_citizenship     155
sub_position                13
position                     4
foot                         3
height_in_cm                48
market_value               223
date                      2141
age                         30
contract_days_left          89
month                       12
year                        12
goals                       51
assists                     31
yellow_cards                23
red_cards                    4
minutes_played            4079
player_club_id             426
league_id                   14
net_transfer_record        272
national_team_players       23
dtype: int64

the `foot`,`position`,`subposition` and `domestic competition id` collumns all have a low enough cardinality so one hot encoding would definitely be suitable.

In [14]:
#for train
encodecols = ["foot", "position", "sub_position", "league_id"]
X_train = pd.get_dummies(X_train, columns=encodecols, prefix=encodecols)

#for val
X_val=pd.get_dummies(X_val,columns=encodecols,prefix=encodecols)

#for test
X_test=pd.get_dummies(X_test,columns=encodecols,prefix=encodecols)

In [15]:
X_train.sample(2)

last_season_x country_of_citizenship  height_in_cm  age  \
63954           2023                Morocco         181.0   31   
14215           2023            Netherlands         196.0   32   

       contract_days_left  month  year  goals  assists  yellow_cards  \
63954                 232     12  2021      1        1             2   
14215                 232      6  2021     15        2             1   

       red_cards  minutes_played  player_club_id  net_transfer_record  \
63954          0             889            1095                  0.0   
14215          0            1968            2282           21850000.0   

       national_team_players  foot_both  foot_left  foot_right  \
63954                      5          0          0           1   
14215                      9          0          0           1   

       position_Attack  position_Defender  position_Goalkeeper  \
63954                0                  1                    0   
14215                1                  0                    0   

       position_Midfield  sub_position_Attacking Midfield  \
63954                  0                                0   
14215                  0                                0   

       sub_position_Central Midfield  sub_position_Centre-Back  \
63954                              0                         0   
14215                              0                         0   

       sub_position_Centre-Forward  sub_position_Defensive Midfield  \
63954                            0                                0   
14215                            1                                0   

       sub_position_Goalkeeper  sub_position_Left Midfield  \
63954                        0                           0   
14215                        0                           0   

       sub_position_Left Winger  sub_position_Left-Back  \
63954                         0                       0   
14215                         0                       0   

       sub_position_Right Midfield  sub_position_Right Winger  \
63954                            0                          0   
14215                            0                          0   

       sub_position_Right-Back  sub_position_Second Striker  league_id_BE1  \
63954                        1                            0              0   
14215                        0                            0              1   

       league_id_DK1  league_id_ES1  league_id_FR1  league_id_GB1  \
63954              0              0              1              0   
14215              0              0              0              0   

       league_id_GR1  league_id_IT1  league_id_L1  league_id_NL1  \
63954              0              0             0              0   
14215              0              0             0              0   

       league_id_PO1  league_id_RU1  league_id_SC1  league_id_TR1  \
63954              0              0              0              0   
14215              0              0              0              0   

       league_id_UKR1  
63954               0  
14215               0

In [16]:
X_test.sample(2)

last_season_x country_of_citizenship  height_in_cm  age  \
13438           2022                Germany         185.0   21   
35990           2023                 France         187.0   23   

       contract_days_left  month  year  goals  assists  yellow_cards  \
13438                 597      6  2023      1        0             0   
35990                1327      6  2023      1        4             4   

       red_cards  minutes_played  player_club_id  net_transfer_record  \
13438          0             241              44            6700000.0   
35990          0            1710           23826           88200000.0   

       national_team_players  foot_both  foot_left  foot_right  \
13438                      7          0          0           1   
35990                     17          0          0           1   

       position_Attack  position_Defender  position_Goalkeeper  \
13438                1                  0                    0   
35990                0                  1                    0   

       position_Midfield  sub_position_Attacking Midfield  \
13438                  0                                0   
35990                  0                                0   

       sub_position_Central Midfield  sub_position_Centre-Back  \
13438                              0                         0   
35990                              0                         1   

       sub_position_Centre-Forward  sub_position_Defensive Midfield  \
13438                            1                                0   
35990                            0                                0   

       sub_position_Goalkeeper  sub_position_Left Midfield  \
13438                        0                           0   
35990                        0                           0   

       sub_position_Left Winger  sub_position_Left-Back  \
13438                         0                       0   
35990                         0                       0   

       sub_position_Right Midfield  sub_position_Right Winger  \
13438                            0                          0   
35990                            0                          0   

       sub_position_Right-Back  sub_position_Second Striker  league_id_BE1  \
13438                        0                            0              0   
35990                        0                            0              0   

       league_id_DK1  league_id_ES1  league_id_FR1  league_id_GB1  \
13438              0              0              0              0   
35990              0              0              0              0   

       league_id_GR1  league_id_IT1  league_id_L1  league_id_NL1  \
13438              0              0             1              0   
35990              0              0             1              0   

       league_id_PO1  league_id_RU1  league_id_SC1  league_id_TR1  \
13438              0              0              0              0   
35990              0              0              0              0   

       league_id_UKR1  
13438               0  
35990               0

In [17]:
X_val.sample(2)

last_season_x country_of_citizenship  height_in_cm  age  \
22114           2023                  Spain         176.0   29   
73159           2023                England         175.0   20   

       contract_days_left  month  year  goals  assists  yellow_cards  \
22114                 232     10  2020      4        4             8   
73159                1327     12  2021      0        0             0   

       red_cards  minutes_played  player_club_id  net_transfer_record  \
22114          0            3150              13           57300000.0   
73159          0              24             405          -60450000.0   

       national_team_players  foot_both  foot_left  foot_right  \
22114                     14          0          0           1   
73159                     14          0          0           1   

       position_Attack  position_Defender  position_Goalkeeper  \
22114                0                  0                    0   
73159                1                  0                    0   

       position_Midfield  sub_position_Attacking Midfield  \
22114                  1                                0   
73159                  0                                0   

       sub_position_Central Midfield  sub_position_Centre-Back  \
22114                              1                         0   
73159                              0                         0   

       sub_position_Centre-Forward  sub_position_Defensive Midfield  \
22114                            0                                0   
73159                            1                                0   

       sub_position_Goalkeeper  sub_position_Left Midfield  \
22114                        0                           0   
73159                        0                           0   

       sub_position_Left Winger  sub_position_Left-Back  \
22114                         0                       0   
73159                         0                       0   

       sub_position_Right Midfield  sub_position_Right Winger  \
22114                            0                          0   
73159                            0                          0   

       sub_position_Right-Back  sub_position_Second Striker  league_id_BE1  \
22114                        0                            0              0   
73159                        0                            0              0   

       league_id_DK1  league_id_ES1  league_id_FR1  league_id_GB1  \
22114              0              1              0              0   
73159              0              0              0              1   

       league_id_GR1  league_id_IT1  league_id_L1  league_id_NL1  \
22114              0              0             0              0   
73159              0              0             0              0   

       league_id_PO1  league_id_RU1  league_id_SC1  league_id_TR1  \
22114              0              0              0              0   
73159              0              0              0              0   

       league_id_UKR1  
22114               0  
73159               0

This seems to have worked well.The only collumn i need to deal with now is the `country_of_citizenship` collumn which consists of 155 unique values which on its own is far too many to one hot encode. The likely solution to this would be to get the top N-countries of citizenship in terms of numbers and have one designated collumn named other for every other nation other than them. This seems viable as nationalities in this dataset arent uniformly distributed as seen in our EDA in the previous notebook as the top 10 nationalities accounted for a disproportionate amount of the dataset.

In [18]:
#get the value counts for each nation
country_counts = df["country_of_citizenship"].value_counts()

#get the top 50 and its index
top_countries=country_counts.head(25).index



In [19]:
top_countries

Index(['Spain', 'France', 'Netherlands', 'Brazil', 'Turkey', 'Germany',
       'Portugal', 'Italy', 'Russia', 'England', 'Ukraine', 'Belgium',
       'Denmark', 'Greece', 'Argentina', 'Scotland', 'Serbia', 'Croatia',
       'Senegal', 'Nigeria', 'Sweden', 'Morocco', 'Uruguay', 'Cote d'Ivoire',
       'Poland'],
      dtype='object')

I need to now set everything not in that index to "Other".

In [20]:
#create lambda function setting "Other" if not in index for train,test and val
X_train["country_of_citizenship"]=X_train["country_of_citizenship"].apply(lambda x: x if x in top_countries else 'Other')
X_val["country_of_citizenship"]=X_val["country_of_citizenship"].apply(lambda x: x if x in top_countries else 'Other')
X_test["country_of_citizenship"]=X_test["country_of_citizenship"].apply(lambda x: x if x in top_countries else 'Other')

In [21]:
X_train.head()

last_season_x country_of_citizenship  height_in_cm  age  \
76575            2022                  Italy         184.0   24   
115759           2021                  Spain         180.0   20   
38683            2022                  Other         187.0   26   
44111            2023                  Spain         183.0   18   
90042            2022                  Other         181.0   22   

        contract_days_left  month  year  goals  assists  yellow_cards  \
76575                  597      6  2018      0        0             0   
115759                -134      1  2015      1        0             0   
38683                  232      6  2018      2        0             6   
44111                  597      2  2020      0        0             1   
90042                 1327     10  2017      0        0             4   

        red_cards  minutes_played  player_club_id  net_transfer_record  \
76575           0              23               5          -45000000.0   
115759          0             161             150           43500000.0   
38683           0             810             618           12850000.0   
44111           0             281            1108           -6300000.0   
90042           0            1211             618           12850000.0   

        national_team_players  foot_both  foot_left  foot_right  \
76575                      15          0          0           1   
115759                      7          0          0           1   
38683                       3          0          0           1   
44111                       3          0          1           0   
90042                       3          0          0           1   

        position_Attack  position_Defender  position_Goalkeeper  \
76575                 0                  1                    0   
115759                1                  0                    0   
38683                 0                  0                    0   
44111                 0                  1                    0   
90042                 0                  1                    0   

        position_Midfield  sub_position_Attacking Midfield  \
76575                   0                                0   
115759                  0                                0   
38683                   1                                0   
44111                   0                                0   
90042                   0                                0   

        sub_position_Central Midfield  sub_position_Centre-Back  \
76575                               0                         0   
115759                              0                         0   
38683                               1                         0   
44111                               0                         0   
90042                               0                         0   

        sub_position_Centre-Forward  sub_position_Defensive Midfield  \
76575                             0                                0   
115759                            0                                0   
38683                             0                                0   
44111                             0                                0   
90042                             0                                0   

        sub_position_Goalkeeper  sub_position_Left Midfield  \
76575                         0                           0   
115759                        0                           0   
38683                         0                           0   
44111                         0                           0   
90042                         0                           0   

        sub_position_Left Winger  sub_position_Left-Back  \
76575                          0                       0   
115759                         1                       0   
38683                          0                       0   
44111                          0                       1   
90042                      

I can now perform one hot encoding without excessively increasing dimensionality.

In [22]:


#train
X_train = pd.get_dummies(X_train, columns=["country_of_citizenship"], prefix="Country")

#val
X_val = pd.get_dummies(X_val, columns=["country_of_citizenship"], prefix="Country")

#test
X_test = pd.get_dummies(X_test, columns=["country_of_citizenship"], prefix="Country")


In [23]:
X_train.sample(5)

last_season_x  height_in_cm  age  contract_days_left  month  year  \
2293             2023         181.0   26                 597      5  2019   
94005            2023         184.0   27                 962      3  2022   
114119           2019         187.0   32                -134     12  2020   
36830            2018         197.0   28                 202     10  2015   
50964            2023         185.0   24                 962      7  2020   

        goals  assists  yellow_cards  red_cards  minutes_played  \
2293        1        2             8          1            2635   
94005       0        1             0          0            1080   
114119      0        0             2          0             753   
36830       0        0             1          1            1992   
50964       1        0             2          0            1338   

        player_club_id  net_transfer_record  national_team_players  foot_both  \
2293               265           -7000000.0                     10          0   
94005              468            -650000.0                      2          0   
114119             589            9430000.0                      4          0   
36830              720           34100000.0                      7          0   
50964              157            2520000.0                      8          0   

        foot_left  foot_right  position_Attack  position_Defender  \
2293            0           1                0                  0   
94005           0           1                0                  0   
114119          0           1                0                  1   
36830           0           1                0                  0   
50964           0           1                0                  1   

        position_Goalkeeper  position_Midfield  \
2293                      0                  1   
94005                     1                  0   
114119                    0                  0   
36830                     1                  0   
50964                     0                  0   

        sub_position_Attacking Midfield  sub_position_Central Midfield  \
2293                                  0                              0   
94005                                 0                              0   
114119                                0                              0   
36830                                 0                              0   
50964                                 0                              0   

        sub_position_Centre-Back  sub_position_Centre-Forward  \
2293                           0                            0   
94005                          0                            0   
114119                         1                            0   
36830                          0                            0   
50964                          1                            0   

        sub_position_Defensive Midfield  sub_position_Goalkeeper  \
2293                                  1                        0   
94005                                 0                        1   
114119                                0                        0   
36830                                 0                        1   
50964                                 0                        0   

        sub_position_Left Midfield  sub_position_Left Winger  \
2293                             0                         0   
94005                            0                         0   
114119                           0                         0   
36830                            0                         0   
50964                            0                         0   

        sub_position_Left-Back  sub_position_Right Midfield  \
2293                         0                            0   
94005                        0                            0   
114119                       0                            0   
36830                        0                           

I've now completed my feature encoding and can move onto the next step of my preprocessing.

## Scaling Data

I'll use the StandardScaler as there will be quite a few outliers as bigger teams in better leagues have a disproportionate amount of funds to spend on players which min-max scaling is very sensitive to.

In [24]:
from sklearn.preprocessing import StandardScaler
#instantiate
scaler=StandardScaler()

#fit and transform
#train
X_train_s=scaler.fit_transform(X_train)

#val
X_val_s=scaler.transform(X_val)


#test
X_test_s=scaler.transform(X_test)




They're now in array form and I want it back in dataframe format.

In [25]:
#train
X_train_s = pd.DataFrame(X_train_s, columns=X_train.columns)

#val
X_val_s = pd.DataFrame(X_val_s, columns=X_val.columns)


#test
X_test_s = pd.DataFrame(X_test_s, columns=X_test.columns)



In [26]:
X_train_s.sample(10)

last_season_x  height_in_cm       age  contract_days_left     month  \
41815      -1.923271      0.709356  0.240506           -1.223367 -1.511990   
54035       0.686577      0.559756 -0.511092            1.872191  0.718179   
26869       0.251602     -1.085852  2.996366           -0.447889  1.275721   
81154      -1.053322      1.906162  1.493170           -1.223367  1.554492   
46079      -1.053322      0.709356  0.491039           -1.223367 -0.118134   
13221       0.686577      1.457360 -0.761625            1.098831 -0.118134   
27718       0.686577      1.008558 -0.511092            1.098831  1.554492   
48679      -0.183373     -0.188248  1.743703           -1.223367 -0.118134   
19503       0.686577     -0.038647 -1.262690            0.325471 -0.118134   
31533      -0.618347     -1.085852 -0.010027           -1.223367 -1.233219   

           year     goals   assists  yellow_cards  red_cards  minutes_played  \
41815 -0.230274 -0.532255 -0.205736      0.445709  -0.258501       -0.187111   
54035  1.113205  0.547121  0.201472     -0.960436  -0.258501        1.103770   
26869  1.113205 -0.532255  0.201472      1.500317  -0.258501        0.227001   
81154  0.105595 -0.532255  0.201472     -0.257364  -0.258501        1.142408   
46079 -1.909624 -0.262411 -0.612944     -0.608900  -0.258501       -0.971745   
13221 -0.566144 -0.532255 -0.205736     -0.960436  -0.258501       -0.670572   
27718  0.777335 -0.532255 -0.612944      0.797245  -0.258501       -0.088041   
48679 -0.230274  0.007433 -0.205736      1.148781  -0.258501        0.434058   
19503 -0.566144 -0.532255 -0.612944     -0.960436  -0.258501       -1.217438   
31533 -0.902014  1.626497  1.015889     -0.257364  -0.258501        1.115659   

       player_club_id  net_transfer_record  national_team_players  foot_both  \
41815       -0.323620            -0.002528              -0.673275  -0.203311   
54035       -0.305559            -2.842402               1.026196  -0.203311   
26869       -0.019541            -0.002528              -1.050935  -0.203311   
81154       -0.036705            -0.002528              -0.862105  -0.203311   
46079       -0.388560            -0.381701               1.403856  -0.203311   
13221       -0.334891            -0.193422               1.026196  -0.203311   
27718       -0.392915            -0.066595              -0.673275   4.918564   
48679       -0.349877            -0.875933               1.403856   4.918564   
19503       -0.323620            -0.002528              -0.673275  -0.203311   
31533       -0.354104            -0.094837               0.648535  -0.203311   

       foot_left  foot_right  position_Attack  position_Defender  \
41815   1.731702   -1.565570        -0.635055           1.407452   
54035  -0.577466    0.638745        -0.635055          -0.710504   
26869  -0.577466    0.638745        -0.635055          -0.710504   
81154  -0.577466    0.638745        -0.635055           1.407452   
46079  -0.577466    0.638745         1.574668          -0.710504   
13221  -0.577466    0.638745        -0.635055           1.407452   
27718  -0.577466   -1.565570        -0.635055           1.407452   
48679  -0.577466   -1.565570        -0.635055           1.407452   
19503   1.731702   -1.565570         1.574668          -0.710504   
31533  -0.577466    0.638745        -0.635055          -0.710504   

       position_Goalkeeper  position_Midfield  \
41815            -0.297908          -0.647845   
54035            -0.297908           1.543580   
26869            -0.297908           1.543580   
81154            -0.297908          -0.647845   
46079            -0.297908          -0.647845   
13221            -0.297908          -0.647845   
27718            -0.297908          -0.647845   
48679            -0.297908          -0.647845   
19503            -0.297908          -0.647845   
31533            -0.297908           1.543580   

       sub_position_Attacking Midfield  sub_position_Central Midfield  \
41815                   

In [27]:
#single out Cristiano Ronaldo using scaled data
X_train_s.loc[
    (X_train_s['position_Attack'] > 1.0) & (X_train_s['position_Attack'] < 2.0) &
    (X_train_s['Country_Portugal'] > 4.0) & (X_train_s['Country_Portugal'] < 5.0) &
    (X_train_s['foot_right'] > -1.6) & (X_train_s['foot_right'] < -1.5)&
    (X_train_s["height_in_cm"]>0.7) & (X_train_s["height_in_cm"]<0.71)
]



last_season_x  height_in_cm       age  contract_days_left     month  \
1122        0.686577      0.709356 -0.761625            1.098831  0.160637   
3336        0.686577      0.709356 -0.010027            1.098831 -0.118134   
10245       0.686577      0.709356 -0.761625            1.098831 -1.233219   
14562       0.686577      0.709356 -0.010027            1.098831  1.275721   
26416       0.686577      0.709356  1.493170            1.098831 -0.118134   
27580       0.686577      0.709356  0.491039            1.098831 -0.118134   
40603       0.686577      0.709356  0.741572            1.098831  0.439408   
55824       0.686577      0.709356  0.992104            1.098831 -0.396905   
58852       0.686577      0.709356  1.242637            1.098831 -0.396905   
69160       0.686577      0.709356  1.242637            1.098831 -0.954448   
76847       0.686577      0.709356  0.741572            1.098831 -1.511990   
80396       0.686577      0.709356  0.240506            1.098831 -0.118134   
82494       0.686577      0.709356  0.491039            1.098831 -0.675677   

           year     goals   assists  yellow_cards  red_cards  minutes_played  \
1122  -1.573754 -0.262411 -0.612944      0.445709  -0.258501       -0.669582   
3336  -0.566144  1.356653  0.608680      0.094172  -0.258501        0.156662   
10245 -1.237884 -0.532255 -0.612944     -0.608900  -0.258501       -1.047038   
14562 -0.566144  1.356653  0.608680      0.094172  -0.258501        0.156662   
26416  1.449075  2.436029  2.237513      0.797245  -0.258501        0.490527   
27580  0.105595  3.515405  0.608680      0.797245  -0.258501        0.956156   
40603  0.441465  5.134469  3.459137      1.500317  -0.258501        1.822028   
55824  0.777335  2.705873  2.644721      2.203390  -0.258501        1.802214   
58852  1.113205  3.245561  2.644721      1.851853  -0.258501        1.121603   
69160  1.449075  2.436029  2.237513      0.797245  -0.258501        0.490527   
76847  0.777335  2.705873  2.644721      2.203390  -0.258501        1.802214   
80396 -0.230274  3.515405  1.423097      0.094172  -0.258501        0.811514   
82494  0.441465  5.134469  3.459137      1.500317  -0.258501        1.822028   

       player_club_id  net_transfer_record  national_team_players  foot_both  \
1122        -0.092679             0.204056              -0.673275  -0.203311   
3336        -0.265468            -0.177732              -0.106785  -0.203311   
10245       -0.092679             0.204056              -0.673275  -0.203311   
14562       -0.265468            -0.177732              -0.106785  -0.203311   
26416       -0.360124             1.814888               0.270875  -0.203311   
27580       -0.265468            -0.177732              -0.106785  -0.203311   
40603       -0.265468            -0.177732              -0.106785  -0.203311   
55824       -0.265468            -0.177732              -0.106785  -0.203311   
58852       -0.360124             1.814888               0.270875  -0.203311   
69160       -0.360124             1.814888               0.270875  -0.203311   
76847       -0.265468            -0.177732              -0.106785  -0.203311   
80396       -0.265468            -0.177732              -0.106785  -0.203311   
82494       -0.265468            -0.177732              -0.106785  -0.203311   

       foot_left  foot_right  position_Attack  position_Defender  \
1122    1.731702    -1.56557         1.574668          -0.710504   
3336    1.731702    -1.56557         1.574668          -0.710504   
10245   1.731702    -1.56557         1.574668          -0.710504   
14562   1.731702    -1.56557         1.574668          -0.710504   
26416   1.731702    -1.56557         1.574668          -0.710504   
27580   1.731702    -1.56557         1.574668          -0.710504   
40603   1.731702    -1.56557         1.574668          -0.710504   
55824   1.731702    -1.56557         1.574668          -0.710504   
58852   1.731702    -1.56557         1.574668          -0.71

Data looks correctly scaled. We can now move onto our Baseline linear Model.

### Data Export

In [29]:
import pickle
df.to_pickle("data1/df.pkl")
X_train_s.to_pickle('data1/X_train_s.pkl')
X_val_s.to_pickle('data1/X_val_s.pkl')
X_test_s.to_pickle('data1/X_test_s.pkl')
y_train.to_pickle('data1/y_train.pkl')
y_val.to_pickle('data1/y_val.pkl')
y_test.to_pickle('data1/y_test.pkl')

In [29]:
import joblib

# Save the trained model to a file
scaler_filename = 'myscaler.joblib'
joblib.dump(scaler, scaler_filename)

print(f"Scaler saved to {scaler_filename}")


Scaler saved to myscaler.joblib
